astropy/table/serialize.py

```diff
@@ -0,0 +1,165 @@
+                                            ('description', lambda x: x is not None, None),
+                                            ('meta', lambda x: x, None)):
+                col_attr = getattr(col.info, attr)
+                if nontrivial(col_attr):
+                    info[attr] = xform(col_attr) if xform else col_attr
+
+            for data_attr in data_attrs:
+                data = obj_attrs[data_attr]
+
+                name = col.info.name
+                # New column name is the same as original if there is only one
+                # data attribute (e.g. time.value), otherwise make a new name
+                # (e.g. skycoord.ra, skycoord.dec).
+                #
+                # In the single-name case the mixin Info gets put into the new
+                # Column so the ECSV header looks complete (and for back compatibility
+                # with pre-2.0 behavior).  However, these are ignored in object
+                # construction later in favor of values in the
+                # __mixin_columns__['__info__'] dict.  This is because in general
+                # object constructors don't pay attention to the .info attributes
+                # in the input data args (e.g. Quantity(col) does look at the unit
+                # but not info.description).
+                if len(data_attrs) > 1:
+                    name = name + '.' + data_attr
+                    col_info = {}
+                else:
+                    col_info = info
+                new_cols.append(Column(data, name=name, **col_info))  # TODO masking, MaskedColumn
+
+                obj_attrs[data_attr] = SerializedColumn({'name': name})
+
+            # Store the fully qualified class name
+            obj_attrs['__class__'] = col.__module__ + '.' + col.__class__.__name__
+
+            if info:
+                # Strip out any attributes defined by the parent
+                for attr in col.info.attrs_from_parent:
+                    if attr in info:
+                        del info[attr]
+                obj_attrs['__info__'] = info
+
+            mixin_cols[col.info.name] = obj_attrs
+
+    meta = deepcopy(tbl.meta)
+    meta['__mixin_columns__'] = mixin_cols
+    out = Table(new_cols, meta=meta, copy=False)
+
+    return out
+
+
+def _construct_mixin_from_obj_attrs(obj_attrs):
+    cls_full_name = obj_attrs.pop('__class__')
+
+    # If this is a supported class then import the class and run
+    # the _construct_from_col method.  Prevent accidentally running
+    # untrusted code by only importing known astropy classes.
+    if cls_full_name not in __construct_mixin_classes:
+        raise ValueError('unsupported class for construct {}'.format(cls_full_name))
+
+    mod_name, cls_name = re.match(r'(.+)\.(\w+)', cls_full_name).groups()
+    module = import_module(mod_name)
+    cls = getattr(module, cls_name)
+    return cls.info._construct_from_dict(obj_attrs)
```

Comment:
I guess one solution to the constructor not setting `info` attributes would be to change this `_construct_from_dict` to also take `info_attrs`. It might be as simple as making a new method `MixinInfo._set_info_attributes()`? Though just doing it explicitly below is fine for now too.